packet_link_qualification/packet_link_qualification.proto

```diff
@@ -6,207 +6,365 @@ syntax = "proto3";
-  // The IDs assigned to the given qualification operations.
-  repeated string ids = 1;
+message AsicLoopbackConfiguration {
+  // This is where any l2/l3/l4 match criteria would be described.
+}
+// GetRequest returns the status for the provided ids.
+message GetRequest{
+  repeated string ids =1;
 }
 
-message StopPacketQualificationResponse {
-  // Per interface stop response.
-  message StopResponse {
-    // The same ID given in the stop request.
-    string id = 1;
-    // Status of the operation. In case of failures, canonical error code and
-    // message will show the details.
-    google.rpc.Status status = 2;
-  }
+// GetResponse returns a map containing the values for all requested
+// Qualification results. If the QualificationResult state is
+// QUALIFICATION_STATE_ERROR the caller should inspect the status field for
+// the exact error code and message.
+// Expected errors codes:
+// NOT_FOUND when the requested id was not found by the service.
+// INVALID_ARGUMENT for any configuration parameter which is unsupported.
+message GetResponse{
+  map<string, QualificationResult> results = 1;
+}
+
+// States of qualification.
+enum QualificationState {
+  QUALIFICATION_STATE_UNSPECIFIED = 0;
+  QUALIFICATION_STATE_ERROR = 1;  // The qualification has errored.
+  QUALIFICATION_STATE_IDLE = 2;       // Initial state for the qualification.
+  QUALIFICATION_STATE_SETUP = 3;       // Interface is being configured.
+  QUALIFICATION_STATE_RUNNING = 4;    // Qualification underway.
+  QUALIFICATION_STATE_TEARDOWN = 5;    // Interface is being reset.
+  QUALIFICATION_STATE_COMPLETED = 6;  // Qualification is complete.
+}
+
+message QualificationResult {
+  // ID of the qualification.
+  string id = 1;
+
+  // Interface name of the qualification.
+  string interface_name = 2;
 
-  repeated StopResponse responses = 1;
+  // The state the test was in when the results were snapshotted.
+  QualificationState state = 3;
+
+  // The number of qualification packets sent.
+  uint64 packets_sent = 4;
+
+  // The number of qualification packets received.
+  uint64 packets_received = 5;
+
+  // The number of packets transmitted that experienced corruption.
+  uint64 packets_error = 6;
+
+  // The number of packets dropped by the device due to internal drop,
+  // lookup or forwarding errors.
+  uint64 packets_dropped = 7;
+
+  // The qualification start time. Also when the first snapshot of
+  // results are taken.
+  google.protobuf.Timestamp start_time = 8;
+
+  // The qualification end time or the current snapshot time since epoch.
+  google.protobuf.Timestamp end_time = 9;
+
+  // Expected rate for the qualification. This is the computed or
+  // observed rate that the service expected to be maintained
+  // throughout the qualification duration.
+  uint64 expected_rate_bytes_per_second = 10;
+
+  // The qualification rate observed during the qualification. 
+  // It is updated every update_interval in bytes per second.
```

Comment:
I think this is the first reference to `update_interval`, which is referenced elsewhere. Should this be included in a `CreateRequest`? (It's not in the proto either.)

It seems like `update_interval` might be something that is specific to an underlying device though, so I'm not sure whether it's something that we want the client to specify.

Reply:
updated